lib/Dialect/RTG/Transforms/ElaborationPass.cpp

```diff
@@ -764,21 +775,87 @@ class Elaborator : public RTGOpVisitor<Elaborator, FailureOr<DeletionKind>> {
     return DeletionKind::Delete;
   }
 
-  FailureOr<DeletionKind> visitOp(BagUniqueSizeOp op) {
+  FailureOr<DeletionKind> visitOp(BagUniqueSizeOp op, VisitorInfo &info) {
     auto size = cast<BagValue>(state.at(op.getBag()))->getBag().size();
     auto sizeAttr = IntegerAttr::get(IndexType::get(op->getContext()), size);
     internalizeResult<AttributeValue>(op.getResult(), sizeAttr);
     return DeletionKind::Delete;
   }
 
-  FailureOr<DeletionKind> visitOp(index::AddOp op) {
+  FailureOr<DeletionKind> visitOp(scf::IfOp op, VisitorInfo &info) {
+    bool cond = cast<BoolValue>(state.at(op.getCondition()))->getBool();
+    if (!info.toElaborate) {
+      info.toElaborate = cond ? op.thenBlock() : op.elseBlock();
+      return info.toElaborate ? DeletionKind::Keep : DeletionKind::Delete;
+    }
+
+    for (auto [res, out] : llvm::zip(
+             op.getResults(), info.toElaborate->getTerminator()->getOperands()))
+      state[res] = state.at(out);
+
+    info.toElaborate = nullptr;
+    return DeletionKind::Delete;
+  }
+
+  FailureOr<DeletionKind> visitOp(scf::ForOp op, VisitorInfo &info) {
+    auto *lowerBound = dyn_cast<IndexValue>(state.at(op.getLowerBound()));
+    auto *step = dyn_cast<IndexValue>(state.at(op.getStep()));
+    auto *upperBound = dyn_cast<IndexValue>(state.at(op.getUpperBound()));
+
+    if (!lowerBound || !step || !upperBound)
+      return op->emitOpError("can only elaborate index type iterator");
+
+    // The loop never executes the body
+    if (lowerBound->getIndex() >= upperBound->getIndex()) {
+      for (auto [res, iterArg] : llvm::zip(op->getResults(), op.getInitArgs()))
+        state[res] = state.at(iterArg);
+      return DeletionKind::Delete;
+    }
+
+    // First iteration
```

Comment:
You mean actually cloning the MLIR ops? I intentionally tried to write this in a way that does not require any IR cloning/materialization because that's quite expensive. But from a readability perspective that would be better, I agree.